### PR TITLE
OpenAPI Generator

### PR DESCRIPTION
**Pull request by masseelch**
Implements #10

### CODE REVIEW

File: annotation.go [hunk 1]
```diff
@@ -3,58 +3,41 @@ package elk
...
-func Groups(gs ...string) Annotation {
-	return Annotation{Groups: gs}
-}
-
-// MaxDepth returns a max depth annotation.
-func MaxDepth(d uint) Annotation {
-	return Annotation{MaxDepth: d}
+// SchemaAnnotation annotates an entity with metadata for templates.
+type SchemaAnnotation struct {
+	// ExposeCreate defines if a creation handler should be generated.
```
Comment (a8m) [inline annotation.go:13]:
```suggestion
	// CreatePolicy defines if a creation handler should be generated.
```

File: annotation.go [hunk 2]
```diff
@@ -3,58 +3,41 @@ package elk
...
+	CreatePolicy policy.Policy
+	// ExposeRead defines if a read handler should be generated.
+	ReadPolicy policy.Policy
+	// ExposeUpdate defines if an update handler should be generated.
+	UpdatePolicy policy.Policy
+	// ExposeDelete defines if a delete handler should be generated.
+	DeletePolicy policy.Policy
```
Comment (a8m) [inline annotation.go:20]: Fix godoc.

File: annotation.go [hunk 3]
```diff
@@ -99,6 +112,148 @@ func (a *SchemaAnnotation) Decode(o interface{}) error {
...
+		return SchemaAnnotation{
+			CreatePolicy: policy.Expose,
+			ReadPolicy:   policy.Expose,
+			UpdatePolicy: policy.Expose,
+			DeletePolicy: policy.Expose,
+			ListPolicy:   policy.Expose,
+		}
+	}
+	// If a config is given only set those to policy.Expose that are requested.
+	s := SchemaAnnotation{}
```
Comment (a8m) [inline annotation.go:158]:
```suggestion
	var s SchemaAnnotation
```

File: annotation.go [hunk 4]
```diff
@@ -99,6 +112,148 @@ func (a *SchemaAnnotation) Decode(o interface{}) error {
...
+
+type PolicyConfig uint
+
+const (
+	Create PolicyConfig = iota
+	Read
+	Update
+	Delete
+	List
+)
```
Comment (a8m) [inline annotation.go:143]: It feels to me that this type should be named something like `Op`, `Operation` or `HandlerType`, and the `Expose`/`Exclude` options should be named as `PolicyConfig`. Also, why do you need the `policy` package?

File: extension.go [hunk 5]
```diff
@@ -1,34 +1,50 @@
...
 		entc.DefaultExtension
 		easyjsonConfig EasyJsonConfig
 		hooks          []gen.Hook
 		templates      []*gen.Template
+		config         *Config
+		specHooks      []Hook
 	}
 	// ExtensionOption allows managing Extension configuration using functional arguments.
 	ExtensionOption func(*Extension) error
+	// HandlerOption allows managing RestGenerator configuration using function arguments.
```
Comment (a8m) [inline extension.go:31]:
```suggestion
	// HandlerOption allows managing RESTGenerator configuration using function arguments.
```

File: internal/openapi/ent/entc.go [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+				"apiKeySample": {
+					Type: "apiKey",
+					In:   "header",
+					Name: "X-API-KEY",
+				},
+			}),
+			elk.SpecSecurity([]map[string][]string{
+				{"apiKeySample": {}},
+			}),
+			elk.SpecDump(os.Stdout),
```
Comment (a8m) [inline internal/openapi/ent/entc.go:32]: I really like it. Super clean.

File: internal/openapi/ent/schema/category.go [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+// 	a, err := strconv.Atoi(spl[0])
+// 	if err != nil {
+// 		return err
+// 	}
+// 	*c = Currency{
+// 		Amount: a,
+// 		Unit:   Unit(spl[1]),
+// 	}
+// 	return nil
+// }
```
Comment (a8m) [inline internal/openapi/ent/schema/category.go:53]: ?

File: spec/json.go [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+	"fmt"
+	"sort"
+)
+
+type JSONSpec Spec
+
+func (spec Spec) MarshalJSON() ([]byte, error) {
+	return json.Marshal(struct {
+		JSONSpec
+		Version string `json:"openapi"`
```
Comment (a8m) [inline spec/json.go:15]: Nice trick, but you can define it inside the function. 
```suggestion
func (spec Spec) MarshalJSON() ([]byte, error) {
	type JSONSpec Spec
	return json.Marshal(struct {
		JSONSpec
		Version string `json:"openapi"`
```

File: template/http/response.tmpl [hunk 9]
```diff
@@ -35,6 +35,10 @@
 
     {{- range $n, $v := views $ }}
         type (
+            // $n represents the data serialized for the following serialization group combinations:
```
Comment (a8m) [inline template/http/response.tmpl:38]:
```suggestion
            // {{ $n }} represents the data serialized for the following serialization group combinations:
```

File: view.go [hunk 10]
```diff
@@ -8,49 +8,82 @@ import (
...
-func views(g *gen.Graph) (map[string]*View, error) {
-	gss, err := groupCombinations(g)
-	if err != nil {
-		return nil, err
+// newViews returns a map of all views occurring in the given graph. Key is the view's name.
+func newViews(g *gen.Graph) (map[string]*mergedView, error) {
+	// c := &Config{}
+	// if g.Annotations == nil || g.Annotations[c.Name()] == nil {
+	// 	return nil, errElkExtensionNotFound
+	// }
```
Comment (a8m) [inline view.go:44]: delete

**Review by a8m — COMMENTED** [review]
🚀

File: annotation.go (hunk 4 above)
Comment (masseelch) [reply annotation.go:143]: I had some cyclic dependencies before refactoring and forget to dissolve this one. Thanks!

File: internal/openapi/ent/schema/category.go (hunk 7 above)
Comment (masseelch) [reply internal/openapi/ent/schema/category.go:53]: I don't know :man_shrugging:

File: view.go [hunk 11]
```diff
@@ -68,23 +97,23 @@ func (v View) Name() (string, error) {
 	return fmt.Sprintf("%s%dView", v.Node.Name, h.Sum32()), nil
```
Comment (a8m) [inline view.go:97]: 👍

**Review by a8m — APPROVED** [review]

File: internal/simple/ent/entc.go [hunk 12]
```diff
@@ -1,3 +1,4 @@
+//go:build ignore
 // +build ignore
```
[the remark this answers is not in the document]
Comment (masseelch) [reply internal/simple/ent/entc.go:2]: If the user still has go 1.16 it is needed, isn't it?

[the remark this answers is not in the document]
Comment (a8m) [reply internal/simple/ent/entc.go:2]: Right

**Pull request merged**